### PR TITLE
user_role_workflow_manager bugs fixed

### PR DESCRIPTION
**Pull request by ajithandrewj**
## Description
The below issues are fixed
	1	Unexpected Handling of Unknown Permissions in User Role Workflow Manager
	2	The log return is not correct in the case of 'Update the user with new username'.
	3	 Multiple users creation - update response and msg need handling
	4	 Missing Description Length Validation in Role Creation
	5	Unclear Error Message in Role Creation
        6     Inconsistent Handling of User and Role Details in Workflow Manager

## Type of Change
- [ yes ] Bug fix
- [ no ] New feature
- [ no ] Breaking change
- [ no ] Documentation update

## Checklist
- [ yes ] My code follows the style guidelines of this project
- [ yes ] I have performed a self-review of my own code
- [ yes ] I have commented my code, particularly in hard-to-understand areas
- [ no ] I have made corresponding changes to the documentation
- [ yes ] My changes generate no new warnings
- [ yes ] I have added tests that prove my fix is effective or that my feature works
- [ yes ] New and existing unit tests pass locally with my changes
- [ no ] Any dependent changes have been merged and published in downstream modules
- [ yes ] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [ yes ] Tasks are idempotent (can be run multiple times without changing state)
- [ no ] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [ yes ] Playbooks are modular and reusable
- [ no ] Handlers are used for actions that need to run on change

## Documentation
- [ yes ] All options and parameters are documented clearly.
- [ yes ] Examples are provided and tested.
- [ yes ] Notes and limitations are clearly stated.

### CODE REVIEW

File: plugins/modules/user_role_workflow_manager.py [hunk 1]
```diff
@@ -1277,19 +1338,22 @@ def get_diff_merged(self, config):
...
 
                 if update_required_param.get("role_list"):
-                    user_in_have = self.have["current_user_config"]
-                    update_param = update_required_param
-                    update_param["username"] = user_in_have.get("username")
-                    update_param["user_id"] = user_in_have.get("user_id")
-                    user_info_params = self.snake_to_camel_case(update_param)
-                    task_response = self.update_user(user_info_params)
+                    if self.want["username"] not in self.have["current_user_config"]["username"]:
+                        task_response = {"error_message": "Username for an existing User cannot be updated."}
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:1350]: task_response = {"error_message": "Username for an existing user cannot be updated."}

File: plugins/modules/user_role_workflow_manager.py [hunk 2]
```diff
@@ -2975,6 +3083,64 @@ def verify_diff_deleted(self, config):
...
+
+        if self.no_update_role:
+            no_update_role_msg = "Role(s) '{0}' need no update in Cisco Catalyst Center.".format("', '".join(self.no_update_role))
+            result_msg_list.append(no_update_role_msg)
+
+        if self.deleted_role:
+            delete_role_msg = "Role(s) '{0}' deleted successfully from the Cisco Catalyst Center.".format("', '".join(self.deleted_role))
+            result_msg_list.append(delete_role_msg)
+
+        if self.created_user or self.updated_user or self.deleted_user or self.created_role or self.updated_role or self.deleted_role:
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:3135]: Instead, we can simply check with result_msg_list and set the result.. 

(ie)
```
    if result_msg_list:
        self.result["changed"] = True
        self.msg = " ".join(result_msg_list)
    else:
        self.msg = "No changes were made. No user or role actions were performed in Cisco Catalyst Center."

    self.log(self.msg, "INFO")
    self.result["response"] = self.msg
    
     return self
```

File: plugins/modules/user_role_workflow_manager.py [hunk 3]
```diff
@@ -1024,6 +1033,44 @@ def validate_role_parameters(self, role_key, params_list, role_config, role_para
                                              "DEBUG")
                                     self.validate_string_parameter(param, inventory_management[param], error_messages)
 
+    def identify_invalid_params(self, params, mismatches):
+        """
+        Identify and collect invalid parameters from a dictionary or list based on allowed parameters.
+        Args:
+            - params (dict): The dictionary of parameters to be checked. Nested dictionaries or lists are supported.
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:1040]: Looks like you check both dict and list..

`        params (dict | list): The dictionary or list of parameters to be checked. Nested dictionaries or lists are supported.`

File: plugins/modules/user_role_workflow_manager.py [hunk 4]
```diff
@@ -1024,6 +1033,44 @@ def validate_role_parameters(self, role_key, params_list, role_config, role_para
...
+            "monitoring_and_troubleshooting", "monitoring_settings", "troubleshooting_tools", "data_access", "advanced_network_settings",
+            "image_repository", "network_hierarchy", "network_profiles", "network_settings", "virtual_network", "compliance",
+            "eox", "image_update", "inventory_management", "license", "network_telemetry", "pnp", "provision", "device_configuration",
+            "discovery", "network_device", "port_management", "topology", "app_hosting", "bonjour", "stealthwatch", "umbrella",
+            "apis", "bundles", "events", "reports", "group_based_policy", "ip_based_access_control", "security_advisories",
+            "machine_reasoning", "system_management", "audit_log", "event_viewer", "network_reasoner", "remote_device_support",
+            "scheduler", "search", 'role_name', 'description', 'assurance', 'network_analytics', 'network_design', 'network_provision',
+            'network_services', 'platform', 'security', 'system', 'utilities', 'overall'
+        ]
+        self.log("Iterate through the params to find unknown parameters are present or not", "DEBUG")
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:1059]: self.log("Starting to iterate through params to identify unknown parameters.", "DEBUG")

File: plugins/modules/user_role_workflow_manager.py [hunk 5]
```diff
@@ -1024,6 +1033,44 @@ def validate_role_parameters(self, role_key, params_list, role_config, role_para
...
+
+        if isinstance(params, dict):
+            for key, value in params.items():
+                if key not in allowed_params:
+                    mismatches.append(key)
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:1064]: self.log(f"Invalid parameter detected: {key}", "WARNING")
ERROR or WARNING .. you can check it..

File: plugins/modules/user_role_workflow_manager.py [hunk 6]
```diff
@@ -1024,6 +1033,44 @@ def validate_role_parameters(self, role_key, params_list, role_config, role_para
...
+
+                if isinstance(value, dict) or isinstance(value, list):
+                    self.identify_invalid_params(value, mismatches)
+        elif isinstance(params, list):
+            for item in params:
+                self.identify_invalid_params(item, mismatches)
+
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:1071]: if not mismatches:
        self.log("No invalid parameters found.", "INFO")

File: plugins/modules/user_role_workflow_manager.py [hunk 7]
```diff
@@ -891,7 +896,7 @@ def validate_input_yml(self, user_role_details):
             self.status = "failed"
             return self
 
-        if "role_name" in user_role_details[0] and user_role_details[0].get("role_name") is not None:
+        if "role_details" in self.payload.get("config") and "role_name" in user_role_details[0] and user_role_details[0].get("role_name") is not None:
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:899]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py [hunk 8]
```diff
@@ -943,7 +948,11 @@ def validate_input_yml(self, user_role_details):
                 self.status = "success"
                 return self
 
-        self.msg = "Configuration params like 'username' or 'email' or 'role_name' is not available in the playbook"
+        self.msg = (
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:951]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py [hunk 9]
```diff
@@ -1040,16 +1087,34 @@ def valid_role_config_parameters(self, role_config):
...
-        role_name_regex = re.compile(r"^[a-zA-Z0-9._-]{,25}$")
+        role_name_regex = re.compile(r"^[a-zA-Z0-9._-]{1,25}$")
         role_name_regex_msg = "Role names must be 1 to 25 characters long and should contain only letters, numbers, periods, underscores, and hyphens."
-        self.validate_string_field(role_name, role_name_regex,
-                                   "role_name: '{0}' {1}".format(role_name, role_name_regex_msg), error_messages)
 
-        if role_config.get("description"):
-            self.validate_string_parameter("description", role_config["description"], error_messages)
+        if role_name:
+            self.log("role_nameeee: '{0}'".format(role_name))
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1107]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py (hunk 1 above)
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1350]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py (hunk 2 above)
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:3135]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py (hunk 3 above)
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1040]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py (hunk 4 above)
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1059]: self.log("Starting to iterate through params to identify unknown parameters.", "DEBUG")

File: plugins/modules/user_role_workflow_manager.py (hunk 5 above)
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1064]: self.log(f"Invalid parameter detected: {key}", "ERROR")
Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py (hunk 6 above)
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1071]: Updated the code accordingly.

**Review by madhansansel — APPROVED** [review]

**Pull request merged**